cxx/distributions/dirichlet_categorical.hh

```diff
@@ -0,0 +1,62 @@
+// Copyright 2024
+// See LICENSE.txt
+
+#pragma once
+#include "base.hh"
+
+class DirichletCategorical : public Distribution {
+public:
+    double alpha = 1;  // hyperparameter (applies to all categories)
+    std::vector<int> counts;  // counts of observed categories
+    PRNG *prng;
+
+    DirichletCategorical(PRNG *prng, int k) {  // k is number of categories
+        this->prng = prng;
+        counts = std::vector<int>(k, 0);
+    }
+    void incorporate(double x) {
+        assert(x >= 0 && x < counts.size());
+        counts[size_t(x)] += 1;
+    }
+    void unincorporate(double x) {
+        const size_t y = x;
+        assert(y < counts.size());
+        counts[y] -= 1;
+        assert(0 <= counts[y]);
+    }
+    double logp(double x) const {
+        assert(x >= 0 && x < counts.size());
+        const int n = std::accumulate(counts.cbegin(), counts.cend(), 0);
+        const double numer = log(alpha + counts[size_t(x)]);
+        const double denom = log(n + alpha * counts.size());
+        return numer - denom;
+    }
+    double logp_score() const {
+        const size_t k = counts.size();
+        const double a = alpha * k;
+        const int n = std::accumulate(counts.cbegin(), counts.cend(), 0);
+        const double lg = std::transform_reduce(
+            counts.cbegin(), 
+            counts.cend(), 
+            0, 
+            std::plus{},
+            [&](size_t y) -> double {return lgamma(counts[y] + alpha); }
+        );
+        return lgamma(a) - lgamma(a + n) + lg - k * lgamma(alpha);
```

Comment:
I think this may be missing a term, namely $\frac{\Gamma(n + 1)}{\prod_{0 \leq i < k} {\Gamma(y_i + 1)}}$, from https://en.wikipedia.org/wiki/Dirichlet-multinomial_distribution#Dirichlet-multinomial_as_a_compound_distribution. Since $n$ and $y_i$ are integers, this is actually equal to $\frac{n!}{\prod_{0 \leq i < k} {y_i!}}$. The wiki link also gives an alternative formula in terms of the Beta function. (I am not sure which of these various formulas would be most efficient -- for now, I think any of them will do.)

Reply:
@emilyfertig Your implementation does match https://github.com/probcomp/cgpm/blob/56a481829448bddc9cdfebd42f65023287d5b7c7/src/primitives/categorical.py#L150-L155 -- I might just be confused here. Will try to reason about whether we need this extra term.

Reply:
@emilyfertig After discussing with @Joaoloula and Tim, I agree this looks good as is -- we don't need to marginalize over orderings. Sorry for the confusion!

Reply:
Thanks for reasoning that out -- I followed the CGPM implementation and checked it against https://stephentu.github.io/writeups/dirichlet-conjugate-prior.pdf and hadn't thought through marginalization vs. not.